File: abelkhan_cpp/component/hub/hub.cpp
/*
 * qianqians
 * 2020-1-10
 * hub.cpp
 */

#include <boost/uuid/uuid.hpp>
#include <boost/uuid/uuid_generators.hpp>
#include <boost/uuid/uuid_io.hpp>
#include <boost/lexical_cast.hpp>
#include <boost/thread.hpp>

#include "hub_call_hubmodule.h"
#include "center_call_hubmodule.h"
#include "center_call_servermodule.h"
#include "gate_call_hubmodule.h"
#include "dbproxy_call_hubmodule.h"

#include "hub.h"
#include "centerproxy.h"
#include "center_msg_handle.h"
#include "dbproxyproxy.h"
#include "dbproxy_msg_handle.h"
#include "hubsvrmanager.h"
#include "hub_svr_msg_handle.h"
#include "gatemanager.h"
#include "gate_msg_handle.h"
#include "gc_poll.h"

namespace hub{

hub_service::hub_service(std::string config_file_path, std::string config_name) {
	uuid = boost::lexical_cast<std::string>(boost::uuids::random_generator()());

	_config = std::make_shared<config::config>(config_file_path);
	_center_config = _config->get_value_dict("center");
	_root_config = _config;
	_config = _config->get_value_dict(config_name);

	name = _config->get_value_string("hub_name");
}

void hub_service::init() {
	enet_initialize();

	_timerservice = std::make_shared<service::timerservice>();
	close_handle = std::make_shared<closehandle>();
	hubs = std::make_shared<hubsvrmanager>(shared_from_this());

	auto ip = _config->get_value_string("ip");
	auto port = _config->get_value_int("port");
	auto hub_call_hub = std::make_shared<module::hub_call_hub>();
	hub_call_hub->sig_reg_hub.connect(std::bind(hub_msg::reg_hub, hubs, std::placeholders::_1));
	hub_call_hub->sig_reg_hub_sucess.connect(std::bind(hub_msg::reg_hub_sucess));
	hub_call_hub->sig_hub_call_hub_mothed.connect(std::bind(hub_msg::hub_call_hub_mothed, shared_from_this(), std::placeholders::_1, std::placeholders::_2, std::placeholders::_3));
	auto hub_process = std::make_shared<juggle::process>();
	hub_process->reg_module(hub_call_hub);
	_hub_service = std::make_shared<service::enetacceptservice>(ip, port, hub_process);

	_center_process = std::make_shared<juggle::process>();
	_center_service = std::make_shared<service::connectservice>(_center_process);

	auto gate_process = std::make_shared<juggle::process>();
	_gate_service = std::make_shared<service::enetconnectservice>(gate_process);
	gates = std::make_shared<gatemanager>(_gate_service, shared_from_this());
	auto gate_call_hub = std::make_shared<module::gate_call_hub>();
	gate_call_hub->sig_reg_hub_sucess.connect(std::bind(gate_msg::reg_hub_sucess));
	gate_call_hub->sig_client_connect.connect(std::bind(gate_msg::client_connect, gates, std::placeholders::_1));
	gate_call_hub->sig_client_disconnect.connect(std::bind(gate_msg::client_disconnect, gates, std::placeholders::_1));
	gate_call_hub->sig_client_exception.connect(std::bind(gate_msg::client_exception, gates, std::placeholders::_1));
	gate_call_hub->sig_client_call_hub.connect(std::bind(gate_msg::client_call_hub, shared_from_this(), std::placeholders::_1, std::placeholders::_2, std::placeholders::_3, std::placeholders::_4));
	gate_process->reg_module(gate_call_hub);

	_juggleservice = std::make_shared<service::juggleservice>();
	_juggleservice->add_process(hub_process);
	_juggleservice->add_process(_center_process);
	_juggleservice->add_process(gate_process);
}

void hub_service::connect_center() {
	std::cout << "begin on connect center" << std::endl;

	auto ip = _center_config->get_value_string("ip");
	auto port = _center_config->get_value_int("port");
	auto center_ch = _center_service->connect(ip, port);

	_centerproxy = std::make_shared<centerproxy>(center_ch);

	auto center_call_hub = std::make_shared<module::center_call_hub>();
	auto center_call_server = std::make_shared<module::center_call_server>();
	center_call_server->sig_reg_server_sucess.connect(std::bind(center_msg::reg_server_sucess, _centerproxy));
	center_call_server->sig_close_server.connect(std::bind(center_msg::close_server, close_handle));
	center_call_hub->sig_distribute_server_address.connect(std::bind(center_msg::distribute_server_address, shared_from_this(), std::placeholders::_1, std::placeholders::_2, std::placeholders::_3, std::placeholders::_4));
	center_call_hub->sig_reload.connect(std::bind(center_msg::reload, shared_from_this(), std::placeholders::_1));
	_center_process->reg_module(center_call_hub);
	_center_process->reg_module(center_call_server);

	_centerproxy->reg_server(_config->get_value_string("ip"), _config->get_value_int("port"), uuid);

	std::cout << "end on connect center" << std::endl;
}

void hub_service::connect_gate(std::string uuid, std::string ip, uint16_t port) {
	gates->connect_gate(uuid, ip, port);
}

void hub_service::reg_hub(std::string hub_ip, uint16_t hub_port) {
	_hub_service->connect(hub_ip, hub_port, [this](std::shared_ptr<juggle::Ichannel> ch){
		auto caller = std::make_shared< caller::hub_call_hub>(ch);
		caller->reg_hub(name);
	});
}

void hub_service::try_connect_db(std::string dbproxy_ip, uint16_t dbproxy_port) {
	if (!_config->has_key("dbproxy")) {
		return;
	}

	auto dbproxy_cfg = _root_config->get_value_dict(_config->get_value_string("dbproxy"));
	auto _db_ip = dbproxy_cfg->get_value_string("ip");
	auto _db_port = dbproxy_cfg->get_value_int("port");
	if (dbproxy_ip != _db_ip || dbproxy_port != _db_port) {
		return;
	}

	auto dbproxy_process = std::make_shared<juggle::process>();
	_dbproxy_service = std::make_shared<service::connectservice>(dbproxy_process);
	auto ch = _dbproxy_service->connect(_db_ip, _db_port);
	_dbproxyproxy = std::make_shared<dbproxyproxy>(ch);

	auto dbproxy_call_hub = std::make_shared<module::dbproxy_call_hub>();
	dbproxy_call_hub->sig_reg_hub_sucess.connect(std::bind(db_msg::reg_hub_sucess));
	dbproxy_call_hub->sig_ack_create_persisted_object.connect(std::bind(db_msg::ack_create_persisted_object, _dbproxyproxy, std::placeholders::_1, std::placeholders::_2));
	dbproxy_call_hub->sig_ack_updata_persisted_object.connect(std::bind(db_msg::ack_updata_persisted_objec, _dbproxyproxy, std::placeholders::_1));
	dbproxy_call_hub->sig_ack_get_object_count.connect(std::bind(db_msg::ack_get_object_count, _dbproxyproxy, std::placeholders::_1, std::placeholders::_2));
	dbproxy_call_hub->sig_ack_get_object_info.connect(std::bind(db_msg::ack_get_object_info, _dbproxyproxy, std::placeholders::_1, std::placeholders::_2));
	dbproxy_call_hub->sig_ack_get_object_info_end.connect(std::bind(db_msg::ack_get_object_info_end, _dbproxyproxy, std::placeholders::_1));
	dbproxy_call_hub->sig_ack_remove_object.connect(std::bind(db_msg::ack_remove_object, _dbproxyproxy, std::placeholders::_1));

	dbproxy_process->reg_module(dbproxy_call_hub);
	_juggleservice->add_process(dbproxy_process);
}

void hub_service::poll() {
	auto time_now = msec_time();
	while (1) {
		try {
			_hub_service->poll();
			_center_service->poll();
			_gate_service->poll();
			_dbproxy_service->poll();

			_timerservice->poll();

			_juggleservice->poll();
		}
		catch (std::exception err) {
			std::cout << "error:" << err.what() << std::endl;
		}

		if (close_handle->is_closed) {
			_timerservice->addticktimer(2000, [](uint64_t timetmp) {
				exit(0);
			});
		}
	}
}

}

File: abelkhan_js/component/hub/hub.js
function hub(argvs){
    event_closure.call(this);

    const uuidv1 = require('uuid/v1');
    this.uuid = uuidv1();

    var cfg = config(argvs[0]);
    this.center_cfg = cfg["center"];
    if (argvs.length > 1){
        this.cfg = cfg[argvs[1]];
    }
    this.root_cfg = cfg;

    this.name = this.cfg["hub_name"];

    var path = require('path');
    configLogger(path.join(this.cfg["log_dir"], this.cfg["log_file"]), this.cfg["log_level"]);
    getLogger().trace("config logger!");

    enet.enet_initialize();

    this.modules = new modulemng();

    this.close_handle = new closehandle();
    this.hubs = new hubmng(this);

    var _hub_msg_handle = new hub_msg_handle(this.modules, this.hubs);
    var hub_call_hub = new hub_call_hub_module();
    hub_call_hub.add_event_listen("reg_hub", _hub_msg_handle, _hub_msg_handle.reg_hub);
    hub_call_hub.add_event_listen("reg_hub_sucess", _hub_msg_handle, _hub_msg_handle.reg_hub_sucess);
    hub_call_hub.add_event_listen("hub_call_hub_mothed", _hub_msg_handle, _hub_msg_handle.hub_call_hub_mothed);
    this.hub_process = new juggle_process();
    this.hub_process.reg_module(hub_call_hub);
    this.hub_service = new enetservice(this.cfg["ip"], this.cfg["port"], this.hub_process);
    //this.hub_service = new kcpservice(this.cfg["ip"], this.cfg["port"], this.hub_process);

    this.center_process = new juggle_process();
    this.connect_center_service = new connectservice(this.center_process);
    this.connect_center = ()=>{
        this.connect_center_service.connect(this.center_cfg["ip"], this.center_cfg["port"], this, function(center_ch){
            getLogger().trace("begin on connect center");

            this.centerproxy = new centerproxy(center_ch);

            var center_call_hub = new center_call_hub_module();
            var center_call_server = new center_call_server_module();
            var _center_msg_handle = new center_msg_handle(this, this.centerproxy);
            center_call_server.add_event_listen("reg_server_sucess", _center_msg_handle, _center_msg_handle.reg_server_sucess);
            center_call_server.add_event_listen("close_server", _center_msg_handle, _center_msg_handle.close_server);
            center_call_hub.add_event_listen("distribute_server_address", _center_msg_handle, _center_msg_handle.distribute_server_address);
            center_call_hub.add_event_listen("reload", this, this.onReload_event);
            this.center_process.reg_module(center_call_hub);
            this.center_process.reg_module(center_call_server);

            this.centerproxy.reg_hub(this.cfg["ip"], this.cfg["port"], this.uuid);

            getLogger().trace("end on connect center");
        });
    }

    var gate_call_hub = new gate_call_hub_module();
    var _gate_msg_handle = new gate_msg_handle(this, this.modules);
    gate_call_hub.add_event_listen("reg_hub_sucess", _gate_msg_handle, _gate_msg_handle.reg_hub_sucess);
    gate_call_hub.add_event_listen("client_connect", _gate_msg_handle, _gate_msg_handle.client_connect);
    gate_call_hub.add_event_listen("client_disconnect", _gate_msg_handle, _gate_msg_handle.client_disconnect);
    gate_call_hub.add_event_listen("client_exception", _gate_msg_handle, _gate_msg_handle.client_exception);
    gate_call_hub.add_event_listen("client_call_hub", _gate_msg_handle, _gate_msg_handle.client_call_hub);
    this.gate_process = new juggle_process();
    this.gate_process.reg_module (gate_call_hub);
    this.connect_gate_servcie = this.hub_service;
    //this.connect_gate_servcie = new kcpservice("127.0.0.1", 0, this.gate_process);
    this.gates = new gatemng(this.connect_gate_servcie, this);

    this.juggle_service = new juggleservice();

    if (this.cfg["out_ip"] && this.cfg["out_port"]){
        let xor_key = this.cfg["key"];

        var _direct_client_msg_handle = new direct_client_msg_handle(this.modules, this);
        var client_call_hub = new client_call_hub_module();
        client_call_hub.add_event_listen("client_connect", _direct_client_msg_handle, _direct_client_msg_handle.client_connect);
        client_call_hub.add_event_listen("call_hub", _direct_client_msg_handle, _direct_client_msg_handle.call_hub);
        this.direct_client_process = new juggle_process();
        this.direct_client_process.reg_module(client_call_hub);
        this.accept_client_service = new websocketacceptservice(this.cfg["out_ip"], this.cfg["out_port"], this.direct_client_process);
        var that = this;
        this.accept_client_service.add_event_listen("on_channel_connect", this, (ch)=>{
            ch.xor_key = xor_key % 256;
        });
        this.accept_client_service.add_event_listen("on_channel_disconnect", this, (ch)=>{
            delete that.gates.direct_ch[ch.client_uuid];
        });

        this.juggle_service.add_process(this.direct_client_process);
    }

    this.juggle_service.add_process(this.hub_process);
	this.juggle_service.add_process(this.center_process);
    this.juggle_service.add_process (this.gate_process);

    this.is_busy = false;
    var juggle_service = this.juggle_service;
    var that = this;
    var time_now = Date.now();
    this.poll = ()=>{
        try {
            this.hub_service.poll();
            this.connect_gate_servcie.poll();
            juggle_service.poll();

            that.call_event("on_tick", []);
        }
        catch(err) {
            getLogger().error(err);
        }

        if (that.close_handle.is_close){
            setTimeout(()=>{
                enet.enet_deinitialize();
                process.exit();
            }, 2000);
        }else{
            var _tmp_now = Date.now();
            var _tmp_time = _tmp_now - time_now;
            time_now = _tmp_now;
            if (_tmp_time < 50){
                that.is_busy = false;
                setTimeout(that.poll, 5);
            }
            else{
                that.is_busy = true;
                setImmediate(that.poll);
            }
        }
    }

    this.onConnectDB_event = function(){
        this.call_event("on_connect_db", []);
    }

    this.onConnectCenter_event = function(){
        this.call_event("on_connect_center", []);
    }

    this.onCloseServer_event = function(){
        this.call_event("on_close", []);
        this.centerproxy.closed();
        this.close_handle.is_close = true;
    }

    this.onReload_event = function(argv){
        this.call_event("on_reload", [argv]);
    }

    this.reg_hub = (hub_ip, hub_port)=>{
        this.hub_service.connect(hub_ip, hub_port, (ch)=>{
            var caller = new hub_call_hub_caller(ch);
            caller.reg_hub(this.name);
        });
    };

    this.try_connect_db = (dbproxy_ip, dbproxy_port)=>{
        if (!this.cfg["dbproxy"]){
            return;
        }

        this.dbproxy_cfg = this.root_cfg[this.cfg["dbproxy"]];
        if (dbproxy_ip != this.dbproxy_cfg["ip"] || dbproxy_port != this.dbproxy_cfg["port"]){
            return;
        }

        var dbproxy_call_hub = new dbproxy_call_hub_module();
        var _dbproxy_msg_handle = new dbproxy_msg_handle(this);
        dbproxy_call_hub.add_event_listen("reg_hub_sucess", _dbproxy_msg_handle, _dbproxy_msg_handle.reg_hub_sucess);
        dbproxy_call_hub.add_event_listen("ack_create_persisted_object", _dbproxy_msg_handle, _dbproxy_msg_handle.ack_create_persisted_object);
        dbproxy_call_hub.add_event_listen("ack_updata_persisted_object", _dbproxy_msg_handle, _dbproxy_msg_handle.ack_updata_persisted_object);
        dbproxy_call_hub.add_event_listen("ack_get_object_count", _dbproxy_msg_handle, _dbproxy_msg_handle.ack_get_object_count);
        dbproxy_call_hub.add_event_listen("ack_get_object_info", _dbproxy_msg_handle, _dbproxy_msg_handle.ack_get_object_info);
        dbproxy_call_hub.add_event_listen("ack_get_object_info_end", _dbproxy_msg_handle, _dbproxy_msg_handle.ack_get_object_info_end);
        dbproxy_call_hub.add_event_listen("ack_remove_object", _dbproxy_msg_handle, _dbproxy_msg_handle.ack_remove_object);
        this.dbproxy_process = new juggle_process();
        this.dbproxy_process.reg_module(dbproxy_call_hub);
        this.connect_dbproxy_service = new connectservice(this.dbproxy_process);
        this.connect_dbproxy_service.connect(this.dbproxy_cfg["ip"], this.dbproxy_cfg["port"], this, function(db_ch){
            this.dbproxy = new dbproxyproxy(db_ch);
            this.dbproxy.reg_hub(that.uuid);
        });

        this.juggle_service.add_process(this.dbproxy_process);
    };
}
module.exports.hub = hub;
module.exports.event_cb = event_closure;
